### PR TITLE
fix: fixes update k8s resource bug.

### LINKED ISSUE #1532: Failed to create new infra
*Description*:
>When we apply eg, then apply proxy, eg will export error ”failed to create new infra“

*Repro steps*:
>1、create gateway
2、create proxy, and get eg logs

*Environment*:
>eg is main branch

*Logs*:
```console
2023-06-15 14:06:47 2023-06-15T06:06:47.107Z    ERROR   runner/runner.go:68     failed to create new infra      {"runner": "infrastructure", "error": "failed to create or update deployment envoy-gateway-system/default-proxy: Operation cannot be fulfilled on deployments.apps \"default-proxy\": the object has been modified; please apply your changes to the latest version and try again", "errorVerbose": "Operation cannot be fulfilled on deployments.apps \"default-proxy\": the object has been modified; please apply your changes to the latest version and try again\nfailed to create or update deployment envoy-gateway-system/default-proxy\ngithub.com/envoyproxy/gateway/internal/infrastructure/kubernetes.(*Infra).createOrUpdate\n\t/build/internal/infrastructure/kubernetes/infra.go:64\ngithub.com/envoyproxy/gateway/internal/infrastructure/kubernetes.(*Infra).CreateOrUpdateProxyInfra\n\t/build/internal/infrastructure/kubernetes/proxy_infra.go:27\ngithub.com/envoyproxy/gateway/internal/infrastructure/runner.(*Runner).subscribeToProxyInfraIR.func1\n\t/build/internal/infrastructure/runner/runner.go:67\ngithub.com/envoyproxy/gateway/internal/message.HandleSubscription[...]\n\t/build/internal/message/watchutil.go:36\ngithub.com/envoyproxy/gateway/internal/infrastructure/runner.(*Runner).subscribeToProxyInfraIR\n\t/build/internal/infrastructure/runner/runner.go:57\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1598"}
```

### PR DESCRIPTION
**Pull request by qicz**
Fixes #1532

### CODE REVIEW

**Review by zirain — CHANGES_REQUESTED** [review]
what's the PR for?

**Comment by qicz** [comment]
> what's the PR for?

pls look pr fixes issue

File: internal/infrastructure/kubernetes/infra_client.go [hunk 1]
```diff
@@ -27,17 +28,16 @@ func (cli *InfraClient) CreateOrUpdate(ctx context.Context, key client.ObjectKey
 		if kerrors.IsNotFound(err) {
 			// Create if it does not exist.
 			if err := cli.Client.Create(ctx, specific); err != nil {
-				return err
+				return fmt.Errorf("for Create %v", err)
```
Comment (zirain) [inline internal/infrastructure/kubernetes/infra_client.go:31]: that's the different with before, how can this fix the issue?

Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go:31]: image

Comment (zirain) [reply internal/infrastructure/kubernetes/infra_client.go:31]: can you explain more about the detail about this edge case?

Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go:31]: xref: https://stackoverflow.com/questions/65987577/kubectl-apply-reports-error-operation-cannot-be-fulfilled-on-serviceaccounts

we need to update existing resources so keep `uuid`.  the `resourceVersion` will update at every editing.

Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go:31]: xref: #1041

Comment (zirain) [reply internal/infrastructure/kubernetes/infra_client.go:31]: should we use retry.OnConflict here?

Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go]: did it

Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go:31]: I  think retryOnConflict is unnecessary. the `resourceVersion` has been removed.

File: internal/infrastructure/kubernetes/infra_client.go [hunk 2]
```diff
@@ -23,26 +25,27 @@ func New(cli client.Client) *InfraClient {
...
-			specific.SetUID(current.GetUID())
-			if err := cli.Client.Update(ctx, specific); err != nil {
-				return err
+		} else {
+			// Since the client.Object does not have a specific Spec field to compare
+			// just perform an update for now.
+			if updateChecker() {
+				specific.SetUID(current.GetUID())
+				if err := cli.Client.Update(ctx, specific); err != nil {
+					return fmt.Errorf("for Update %v", err)
```
Comment (zirain) [inline internal/infrastructure/kubernetes/infra_client.go:42]: will these lines make `RetryOnConflict` not working?

File: internal/infrastructure/kubernetes/infra_client.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ func (cli *InfraClient) CreateOrUpdate(ctx context.Context, key client.ObjectKey
 			if kerrors.IsNotFound(err) {
 				// Create if it does not exist.
 				if err := cli.Client.Create(ctx, specific); err != nil {
-					return fmt.Errorf("for Create %v", err)
+					return errors.Wrap(err, "for Create")
```
Comment (zirain) [inline internal/infrastructure/kubernetes/infra_client.go:33]:
```suggestion
					return errors.Wrap(err, "failed to create resource")
```

File: internal/infrastructure/kubernetes/infra_client.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ func (cli *InfraClient) CreateOrUpdate(ctx context.Context, key client.ObjectKey
 			if updateChecker() {
 				specific.SetUID(current.GetUID())
 				if err := cli.Client.Update(ctx, specific); err != nil {
-					return fmt.Errorf("for Update %v", err)
+					return errors.Wrap(err, "for Update")
```
Comment (zirain) [inline internal/infrastructure/kubernetes/infra_client.go:42]: ditto

File: internal/infrastructure/kubernetes/infra_client.go (hunk 3 above)
Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go:33]: https://github.com/envoyproxy/gateway/blob/cc68246562d6b3b1010415612ccdfcf49172d400/internal/infrastructure/kubernetes/infra.go#L54-L72

the `prefix` has already existed

Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go:33]:
```bash
2023-06-19T19:24:31.902+0800    ERROR   infrastructure  runner/runner.go:69     failed to create new infra      {"runner": "infrastructure", "error": "failed to create or update serviceaccount envoy-gateway-system/envoy-default-eg-63636f73: for Update:..",
```

File: internal/infrastructure/kubernetes/infra_client.go (hunk 2 above)
Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go]: change to `erros.Wrap`

File: internal/infrastructure/kubernetes/infra_client.go [hunk 5]
```diff
@@ -23,26 +25,27 @@ func New(cli client.Client) *InfraClient {
...
-		if updateChecker() {
-			specific.SetResourceVersion(current.GetResourceVersion())
-			specific.SetUID(current.GetUID())
-			if err := cli.Client.Update(ctx, specific); err != nil {
-				return err
+		} else {
+			// Since the client.Object does not have a specific Spec field to compare
+			// just perform an update for now.
+			if updateChecker() {
+				specific.SetUID(current.GetUID())
```
Comment (zirain) [inline internal/infrastructure/kubernetes/infra_client.go:40]: can you add `specific.SetResourceVersion(current.GetResourceVersion())` back?

Comment (qicz) [reply internal/infrastructure/kubernetes/infra_client.go:40]: I am testing the `resourceVersion` is unnecessary. And it will be updated on every editing.

**Review by Xunzhuo — APPROVED** [review]

**Pull request merged**